src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -60,6 +62,36 @@ abbreviation add_char : Type (max u v) := (multiplicative R) →* R'
 
 end add_char_def
 
+section group_structure
+
+universes u v
+
+namespace add_char
+
+open multiplicative
+
+variables {R : Type u} [add_comm_group R] {R' : Type v} [comm_monoid R']
+
+/-- An additive character on a commutative additive group has an inverse. -/
+instance has_inv : has_inv (add_char R R') := ⟨λ ψ, ψ.comp neg_add_monoid_hom.to_multiplicative⟩
```

Comment:
Doesn't this form a diamond with `monoid_hom.has_inv`?

Reply:
No, because `monoid_hom.has_inv` requires the *target* to be a group. See [here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Gauss.20sums/near/291071130).

Reply:
(It would if we use this in a situation where `R'` is a group. But in the applications, `R'` will always be a ring, which is at best a `comm_group_with_zero`.)

Reply:
The problem is, you don't get to choose whether `R'` is a group or not. If I'm working with `multiplicative ℤ →* multiplicative ℤ` in another file, I'll get a diamond via this instance.

This problem would go away if you replace `abbreviation` with `def` above; the problem is you're saying "this is just another spelling for a morphism" but then change your mind to "actually this should have different behavior"; but with `abbreviation`, lean is seeing straight through your spelling change and applying the behavior to the original type.